##### salmon从参考基因组提取序列
    samtools=/public/home/software/opt/bio/software/SAMtools/1.9/bin/samtools
    def extractChromosomeSequence(genome_fasta, coordinate):
        chromosome, start, end, strand = coordinate[0], coordinate[1], coordinate[2], coordinate[3]
        # 查询坐标触顶触底问题均已解决，只会报Truncated sequence错误
        if strand == "+":
            with os.popen(f"{samtools} faidx {genome_fasta} {chromosome}:{start}-{end} 2> /dev/null") as fasta:
                return "".join(fasta.read().split("\n")[1: ]) # 去除输出fasta自带的序列名
        if strand == "-":
            with os.popen(f"{samtools} faidx {genome_fasta} {chromosome}:{start}-{end} --reverse-complement 2> /dev/null") as fasta:
                return "".join(fasta.read().split("\n")[1: ])

##### 物种名提取界门纲目科属种信息
    def specie2lineage(specie):
        specie = specie.replace("_", " ")
        # cmd = f"echo {specie} | taxonkit name2taxid | taxonkit lineage -i 2 | cut -f 3"
        cmd = f"echo {specie} | taxonkit name2taxid | cut -f 2 | taxonkit lineage | taxonkit reformat | cut -f 3"
        lineage = os.popen(cmd).read().strip()
        return lineage

    def taxid2lineage(taxid):
        cmd = f"echo {taxid} | taxonkit lineage | taxonkit reformat | cut -f 3"
        lineage = os.popen(cmd).read().strip()
        return lineage